File: src/ProcessScheduler.java
import java.util.Comparator;
import java.util.List;

public class ProcessScheduler {
    private ProcessManager processManager;
    private int quantum;
    private Process currentProcess;

    public ProcessScheduler(ProcessManager processManager) {
        this.processManager = processManager;
        this.quantum = 2; // Quantum padrão
        this.currentProcess = null;
    }

    public void setQuantum(int quantum) {
        this.quantum = Math.max(1, quantum);
    }

    public Process schedule() {
        // Verificar se há algum processo em execução
        Process runningProcess = processManager.getRunningProcess();

        // Se houver processo em execução
        if (runningProcess != null) {
            // Verificar se o quantum acabou
            if (runningProcess.getQuantumLeft() <= 0 && runningProcess.getType().equals("realtime")) {
                runningProcess.setState("ready");
                currentProcess = null;
            }
            // Se o processo estiver esperando input, não fazemos nada
            else if (runningProcess.getInterpreter().isWaitingForInput()) {
                return runningProcess;
            }
            // Se o processo terminou ou foi bloqueado, já foi tratado pelo sistema
            else if (runningProcess.getState().equals("exit") || runningProcess.getState().equals("blocked")) {
                currentProcess = null;
            }
        }

        // Se não há processo em execução, seleciona o próximo
        if (currentProcess == null) {
            // Primeiro, verifica se há processos de tempo real prontos
            List<Process> realtimeProcesses = processManager.getReadyProcesses().stream()
                    .filter(p -> p.getType().equals("realtime"))
                    .sorted(Comparator
                            .comparing(Process::getPriority)
                            .thenComparing(Process::getArrivalTime))
                    .toList();

            // Se houver processos de tempo real prontos
            if (!realtimeProcesses.isEmpty()) {
                currentProcess = realtimeProcesses.get(0);
                currentProcess.setState("running");
                currentProcess.setQuantumLeft(quantum);
            } else {
                // Senão, verificar se há processos de melhor esforço prontos (FCFS)
                List<Process> besteffortProcesses = processManager.getReadyProcesses().stream()
                        .filter(p -> p.getType().equals("besteffort"))
                        .sorted(Comparator.comparing(Process::getArrivalTime))
                        .toList();

                if (!besteffortProcesses.isEmpty()) {
                    currentProcess = besteffortProcesses.get(0);
                    currentProcess.setState("running");
                    // Processos de melhor esforço não têm quantum limitado
                    currentProcess.setQuantumLeft(Integer.MAX_VALUE);
                }
            }
        }

        // Verificar se um processo de tempo real ficou pronto enquanto um de melhor esforço está executando
        if (currentProcess != null && currentProcess.getType().equals("besteffort")) {
            boolean realtimeReady = processManager.getReadyProcesses().stream()
                    .anyMatch(p -> p.getType().equals("realtime"));

            if (realtimeReady) {
                // Preempção: interrompe o processo de melhor esforço
                currentProcess.setState("ready");
                currentProcess = null;
                // O próximo ciclo de escalonamento selecionará o processo de tempo real
                return schedule();
            }
        }

        return currentProcess;
    }

    public boolean executeCurrentProcess() {
        if (currentProcess == null || !currentProcess.getState().equals("running")) {
            return false;
        }

        try {
            return currentProcess.getInterpreter().executeInstruction();
        } catch (Exception error) {
            System.err.println("Erro na execução do processo " + currentProcess.getId() + ": " + error.getMessage());
            processManager.terminateProcess(currentProcess);
            return false;
        }
    }

    public Process getCurrentProcess() {
        return currentProcess;
    }
}

File: js/processScheduler.js
class ProcessScheduler {
    constructor(processManager) {
        this.processManager = processManager;
        this.quantum = 2; // Quantum padrão
        this.currentProcess = null;
    }

    setQuantum(quantum) {
        this.quantum = Math.max(1, parseInt(quantum) || 2);
    }

    schedule() {
        // Verificar se há algum processo em execução
        const runningProcess = this.processManager.getRunningProcess();

        // Se houver processo em execução
        if (runningProcess) {
            // Verificar se o quantum acabou
            if (runningProcess.quantumLeft <= 0 && runningProcess.type === 'realtime') {
                runningProcess.state = 'ready';
                this.currentProcess = null;
            }
            // Se o processo estiver esperando input, não fazemos nada
            else if (runningProcess.interpreter.waitingForInput) {
                return;
            }
            // Se o processo terminou ou foi bloqueado, já foi tratado pelo sistema
            else if (runningProcess.state === 'exit' || runningProcess.state === 'blocked') {
                this.currentProcess = null;
            }
        }

        // Se não há processo em execução, seleciona o próximo
        if (!this.currentProcess) {
            // Primeiro, verifica se há processos de tempo real prontos
            const realtimeProcesses = this.processManager.getReadyProcesses()
                .filter(p => p.type === 'realtime')
                .sort((a, b) => {
                    // Ordenar por prioridade (0 - alta, 1 - baixa)
                    if (a.priority !== b.priority) {
                        return a.priority - b.priority;
                    }
                    // Em caso de mesma prioridade, ordenar por tempo de chegada
                    return a.arrivalTime - b.arrivalTime;
                });

            // Se houver processos de tempo real prontos
            if (realtimeProcesses.length > 0) {
                this.currentProcess = realtimeProcesses[0];
                this.currentProcess.state = 'running';
                this.currentProcess.quantumLeft = this.quantum;
            } else {
                // Senão, verificar se há processos de melhor esforço prontos (FCFS)
                const besteffortProcesses = this.processManager.getReadyProcesses()
                    .filter(p => p.type === 'besteffort')
                    .sort((a, b) => a.arrivalTime - b.arrivalTime);

                if (besteffortProcesses.length > 0) {
                    this.currentProcess = besteffortProcesses[0];
                    this.currentProcess.state = 'running';
                    // Processos de melhor esforço não têm quantum limitado
                    this.currentProcess.quantumLeft = Infinity;
                }
            }
        }

        // Verificar se um processo de tempo real ficou pronto enquanto um de melhor esforço está executando
        if (this.currentProcess && this.currentProcess.type === 'besteffort') {
            const realtimeReady = this.processManager.getReadyProcesses()
                .some(p => p.type === 'realtime');

            if (realtimeReady) {
                // Preempção: interrompe o processo de melhor esforço
                this.currentProcess.state = 'ready';
                this.currentProcess = null;
                // O próximo ciclo de escalonamento selecionará o processo de tempo real
                this.schedule();
            }
        }

        return this.currentProcess;
    }

    executeCurrentProcess() {
        if (!this.currentProcess || this.currentProcess.state !== 'running') {
            return false;
        }

        try {
            return this.currentProcess.interpreter.executeInstruction();
        } catch (error) {
            console.error(`Erro na execução do processo ${this.currentProcess.id}: ${error.message}`);
            this.processManager.terminateProcess(this.currentProcess);
            return false;
        }
    }
}

